cluster/cluster.go

```diff
@@ -64,93 +77,162 @@ const (
 	ON_MASTER_TO_HOSTS_AND_MASTER
 )
 
-type RemoteOutput struct {
-	Scope     int
-	NumErrors int
-	Stdouts   map[int]string
-	Stderrs   map[int]string
-	Errors    map[int]error
-	CmdStrs   map[int]string
+func scopeIsRemote(scope int) bool {
+	return scope == ON_SEGMENTS || scope == ON_SEGMENTS_AND_MASTER || scope == ON_HOSTS || scope == ON_HOSTS_AND_MASTER
+}
+
+func scopeIncludesMaster(scope int) bool {
+	return scope == ON_SEGMENTS_AND_MASTER || scope == ON_MASTER_TO_SEGMENTS_AND_MASTER || scope == ON_HOSTS_AND_MASTER || scope == ON_MASTER_TO_HOSTS_AND_MASTER
+}
+
+func scopeIsHosts(scope int) bool {
+	return scope == ON_HOSTS || scope == ON_HOSTS_AND_MASTER || scope == ON_MASTER_TO_HOSTS || scope == ON_MASTER_TO_HOSTS_AND_MASTER
 }
 
 /*
- * Base cluster functions
+ * A ShellCommand stores a command to be executed (in both executable and
+ * display form), as well as the results of the command execution and the
+ * necessary information to determine how the command will be or was executed.
+ *
+ * It is assumed that before a caller references Content or Host for a given
+ * command, they will check Scope to ensure that that field is meaningful for
+ * that command.  GenerateCommandList sets Host to "" for per-segment commands
+ * and Content to -2 for per-host commands, just to be safe.
  */
+type ShellCommand struct {
+	Scope         int
+	Content       int
+	Host          string
+	Command       *exec.Cmd
+	CommandString string
+	Stdout        string
+	Stderr        string
+	Error         error
+}
 
-func NewCluster(segConfigs []SegConfig) *Cluster {
-	cluster := Cluster{}
-	cluster.Segments = make(map[int]SegConfig, len(segConfigs))
-	for _, seg := range segConfigs {
-		cluster.ContentIDs = append(cluster.ContentIDs, seg.ContentID)
-		cluster.Segments[seg.ContentID] = seg
+func NewShellCommand(scope int, content int, host string, command []string) ShellCommand {
+	return ShellCommand{
+		Scope:         scope,
+		Content:       content,
+		Host:          host,
+		Command:       exec.Command(command[0], command[1:]...),
+		CommandString: strings.Join(command, " "),
 	}
-	cluster.Executor = &GPDBExecutor{}
-	return &cluster
 }
 
-func (cluster *Cluster) GenerateSegmentSSHCommand(contentID int, generateCommand func(int) string) []string {
-	cmdStr := generateCommand(contentID)
-	if contentID == -1 {
-		return []string{"bash", "-c", cmdStr}
-	}
-	return ConstructSSHCommand(cluster.GetHostForContent(contentID), cmdStr)
+/*
+ * A RemoteOutput is used to make it easier to identify the success or failure
+ * of a cluster command and to display the results to the user.
+ */
+type RemoteOutput struct {
+	Scope          int
+	NumErrors      int
+	Commands       []ShellCommand
```

Comment:
Maybe I'm missing something here, but the new implementation of a ShellCommand turns the storing the commands/outputs/errors from a map (`map[int]string`) to a slice ([]Command). This might cause compatibility issues because it removes the ability to retrieve specific output/cmd/err for a given segment based on contentID. With just a slice, in order to find an output of a specific Segment, a user would have to loop through all ShellCommands and ask what its contentID is. Verifying an output based on the contentID of a segment is something that might not be done a lot in production, but I could imagine tests doing so.

Reply:
You are right that it is more difficult to access specific command outputs per segment with the slice. My recollection of how gpbackup uses it and my guess/hope for how others would use it is that it mostly shouldn't matter when exact segment you are referencing because what matters most is if ANY segments encountered errors or if ANY segments had output, so you will loop through all of them anyways. It will change the format of test objects, but the slice might actually be easier to create than the map.

Reply:
Karen's correct, whenever you're running a command against the whole cluster you shouldn't care about individual segments, just the segments in aggregate.  The only case where you'd really want to get segment information one segment at a time would be to print them out in content order or the like, and moving from a map to a slice means that they're now ordered already.